pkg/plugin/plugin.go

```diff
@@ -0,0 +1,137 @@
+package plugin
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/icinga/icinga-notifications/pkg/rpc"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	MethodGetInfo          = "GetInfo"
+	MethodSetConfig        = "SetConfig"
+	MethodSendNotification = "SendNotification"
+)
+
+type Info struct {
+	Name             string          `json:"display_name"`
+	ConfigAttributes json.RawMessage `json:"config_attrs"`
+}
+
+type Contact struct {
+	FullName  string     `json:"full_name"`
+	Addresses []*Address `json:"addresses"`
+}
+
+type Address struct {
+	Type    string `json:"type"`
+	Address string `json:"address"`
+}
+
+type Incident struct {
+	Id                int64  `json:"id"`
+	ObjectDisplayName string `json:"object_display_name"`
+}
+
+type Event struct {
+	Time     time.Time `json:"time"`
+	URL      string    `json:"url"`
+	Type     string    `json:"type"`
+	Severity string    `json:"severity"`
+	Username string    `json:"username"`
+	Message  string    `json:"message"`
+}
+
+type NotificationRequest struct {
+	Contact       *Contact `json:"contact"`
+	Incident      Incident `json:"incident"`
+	Event         *Event   `json:"event"`
+	IcingaWeb2Url string   `json:"icingaweb2_url"`
+}
```

Comment:
Part of documenting the interface between daemon and channel plugins is to check if the structure makes sense and define a clean interface. As just discussed, that could be improved a bit.

Currently, object attributes are split across `incident` and `event` keys in the JSON. It would make sense to add an extra key `object` that contains all the object related attributes (name, url, tags).

Additionally, the channel doesn't really need the Icinga Web 2 URL. It's only used to generate the incident URL, so it would make more sense to add this as `url` within `incident` instead.